src/libre/re_ast.c

```diff
@@ -49,20 +77,34 @@ free_iter(struct ast_expr *n)
 void
 re_ast_free(struct ast_re *ast)
 {
-	free_iter(ast->expr);
+	re_ast_expr_free(ast->expr);
 	free(ast);
 }
 
+void
+re_ast_expr_free(struct ast_expr *n)
+{
+	free_iter(n);
+}
+
 struct ast_expr *
 re_ast_expr_empty(void)
 {
 	struct ast_expr *res = calloc(1, sizeof(*res));
 	if (res == NULL) { return res; }
 	res->t = AST_EXPR_EMPTY;
-
+	res->flags = RE_AST_FLAG_NULLABLE;
 	return res;
 }
 
+/* Note: this returns a single-instance node, which
+ * other functions should not modify. */
```

Comment:
Why not `const`?

Reply:
Because the const-ness of that particular node wouldn't mix with the non-const-ness of all the other node handling functions. Instead, the node free function skips it, as a special case, and the others just shouldn't modify. There could be more assertions, but it might be hard to express within the type system.